lib/kube/proxy/forwarder.go

```diff
@@ -772,22 +861,42 @@ func (f *Forwarder) authorize(ctx context.Context, actx *authContext) error {
 
 	mfaParams := actx.MFAParams(ap.GetRequireMFAType())
 
+	clusterNotFound := trace.AccessDenied("kubernetes cluster %q not found", actx.kubeCluster)
+	var roleMatchers services.RoleMatchers
+	if actx.kubeResource != nil {
+		clusterNotFound = trace.AccessDenied("%s %q from Kubernetes cluster %q not found", actx.kubeResource.Kind, actx.kubeResource.ClusterResource(), actx.kubeCluster)
+		roleMatchers = services.RoleMatchers{
+			// Append a matcher that validates if the Kubernetes resource is allowed
+			// by the roles that satisfy the Kubernetes Cluster.
+			services.NewKubernetesResourceMatcher(*actx.kubeResource),
+		}
+	}
+
 	// Check authz against the first match.
 	//
 	// We assume that users won't register two identically-named clusters with
 	// mis-matched labels. If they do, expect weirdness.
-	clusterNotFound := trace.AccessDenied("kubernetes cluster %q not found", actx.kubeCluster)
 	for _, s := range servers {
 		ks := s.GetCluster()
 		if ks.GetName() != actx.kubeCluster {
 			continue
 		}
 
-		if err := actx.Checker.CheckAccess(ks, mfaParams); err != nil {
+		if err := actx.Checker.CheckAccess(ks, mfaParams, roleMatchers...); err != nil {
 			return clusterNotFound
 		}
+		// If the user has active Access requests we need to validate that they allow
+		// the kubeResource.
+		// This is required because CheckAccess does not validate the subresource type.
```

Comment:
is there a reason not to do it in `CheckAccess`? It would seem more consistent to do it there. Is it for dependency reasons or intentionally only checking the subresource for certain code paths?

Reply:
I did it for two reasons:
1 - We only verify the subresource if you hit directly a Pod endpoint `/api/:ver/namespaces/:podNamespace/pods/:podName/...`.
If you hit a Pod listing endpoint such as `/api/:ver/namespaces/:podNamespace/pods`, we do not validate the resource directly but we exclude all the resources you shouldn't have access to from the response.
2 - In order to do it in `CheckAccess`, it requires me to pass the `SubResource` as part of the `KubeCluster` in order to retrieve it in `accessChecker.checkAllowedResources`. To fulfill everything, I need to add a `SubResource()` method to the `AccessCheckable` interface - i.e. to all resources.

```go
forwarder.go
for _, s := range servers {
	ks := s.GetCluster()
	if ks.GetName() != actx.kubeCluster {
	   continue
	}
        ksCopy:=ks.Copy()
        if actx.kubeResource != nil{
            ksCopy.SetSubResource(actx.kubeResource)
       } 
       if err := actx.Checker.CheckAccess(ks, mfaParams, roleMatchers...); err != nil {
			return clusterNotFound
	}
}

```

```go

func (a *accessChecker) checkAllowedResources(r AccessCheckable) error {
	if len(a.info.AllowedResourceIDs) == 0 {
		// certificate does not contain a list of specifically allowed
		// resources, only role-based access control is used
		return nil
	}

	// Note: logging in this function only happens in debug mode. This is because
	// adding logging to this function (which is called on every resource returned
	// by the backend) can slow down this function by 50x for large clusters!
	isDebugEnabled, debugf := rbacDebugLogger()
        subResource:=r.GetSubResource()
	for _, resourceID := range a.info.AllowedResourceIDs {
		if resourceID.ClusterName == a.localCluster &&
			(resourceID.Kind == r.GetKind() || (resourceID.Kind == types.KindKubePod && r.GetKind() == types.KindKubernetesCluster)) &&
			resourceID.Name == r.GetName() &&(subResource==nil || subResource.match....) {
			// Allowed to access this resource by resource ID, move on to role checks.
			if isDebugEnabled {
				debugf("Matched allowed resource ID %q", types.ResourceIDToString(resourceID))
			}
			return nil
		}
	}

	if isDebugEnabled {
		allowedResources, err := types.ResourceIDsToString(a.info.AllowedResourceIDs)
		if err != nil {
			return trace.Wrap(err)
		}
		err = trace.AccessDenied("access to %v denied, %q not in allowed resource IDs %s",
			r.GetKind(), r.GetName(), allowedResources)
		debugf("Access denied: %v", err)
		return err
	}
	return trace.AccessDenied("access to %v denied, not in allowed resource IDs", r.GetKind())
}

```

I just found it easier and less intrusive to keep the validation in the forwarder path but I am open to other opinions.

Reply:
I think what you have is okay, I was curious of the reason because the comment makes it sound like `CheckAccess` couldn't do the validation, if it just makes more sense to do it here that should be fine